File: 0070-climbing-stairs/0070-climbing-stairs.java
// we can take only 1 or 2 steps

class Solution {
    static int ref(int n, int[] arr) {
        if(n < 0) return 0;
        if(n == 0) return 1;

        if(arr[n] != -1) return arr[n];

        int one = ref(n - 1, arr);
        int two = ref(n - 2, arr);

        arr[n] = one + two;

        return arr[n];
    }

    public int climbStairs(int n) {
        int[] arr = new int[n + 1];
        Arrays.fill(arr, -1);
        
        return ref(n, arr);
    }
}

File: 0070-climbing-stairs/0070-climbing-stairs.js
var climbStairs = function(n) {
    const arr = new Array(n + 1).fill(-1);

    const ref = (n) => {
        // if (arr[n] !== -1) return arr[n]; THIS LINE SHOULD NOT BE WRITTTEN HERE AS N CAN BE NEGATIVE
        if (n < 0) return 0;
        if (n == 0) return 1;
        if (arr[n] !== -1) return arr[n];
    
        arr[n] = ref(n - 1) + ref(n - 2);
        return arr[n];
    };

    return ref(n);
};
